_posts/2019-08-08-hacking-for-cruise-control.md

```diff
@@ -0,0 +1,244 @@
+      offsets.topic.replication.factor: 1
+      transaction.state.log.replication.factor: 1
+      transaction.state.log.min.isr: 1
+      log.message.format.version: "2.3"
+    storage:
+      type: ephemeral
+  zookeeper:
+    replicas: 3
+    storage:
+      type: ephemeral
+  entityOperator:
+    userOperator: {}
+    topicOperator: {}
+```
+Then applying the Kafka custom resource to the cluster:
+
+```bash
+kubectl apply -f kafka.yaml
+```
+The 'metrics.reporters' field can take a list of comma-separated metric reporter class names. At runtime, Kafka will loop through this list and instantiate these metric reporters to run concurrently.
+
+# Accessing Zookeeper
+
+Cruise Control relies on Zookeeper for the current state and configuration of the Kafka cluster. For estimating partition load, monitoring cluster state, and triggering partition reassignments, Cruise Control needs to be able communicate with Zookeeper directly. By default, communication between Zookeeper and all other Strimzi components must be encrypted and authenticated. However, using a hack by Jakub Scholz, one can get around this requirement by creating a Kubernetes service which reuses Strimzi certificates to pipe unencrypted and unauthenticated traffic to Zookeeper. This allows Cruise Control to communicate with Zookeeper without the need of proper authorization and encryption configuration via Java system properties or Stunnel sidecars.
+
+```
+kubectl apply -f https://gist.githubusercontent.com/scholzj/6cfcf9f63f73b54eaebf60738cfdbfae/raw/068d55ac65e27779f3a5279db96bae03cea70acb/zoo-entrance.yaml
+```
+
+**WARNING**: Allowing unauthenticated access to Zookeeper exposes sensitive Kafka metadata for reading and writing by any user. Nefarious users could exploit this access by altering topic configurations to mess up cluster state or by changing ACL access rules to elevate client privileges for stealing data.
+
+# Deploying Cruise Control
+
+With the Kafka cluster running and the Zookeeper endpoint accessible, Cruise Control can now be started safely and begin sampling metrics and monitoring the cluster. To run on Kubernetes, Cruise Control must be formatted into a container image. One can do this using the following Dockerfile. Note that the Cruise Control configuration file, 'cruisecontrol.properties', must be altered to reference the appropriate Kafka bootstrap and Zookeeper service endpoints. In addition to pointing Cruise Control to the proper service endpoints, the Dockerfile sets up the Cruise Control GUI frontend to be available upon startup.
+
+```Dockerfile
+FROM centos:7
+
+ENV CC_VERSION=2.0.57
+ENV JAVA_HOME=/usr/lib/jvm/java-1.8.0-openjdk
+
+RUN yum -y install git java-1.8.0-openjdk-devel && \
+    yum clean all -y
+
+RUN git clone --branch ${CC_VERSION} https://github.com/linkedin/cruise-control.git
+
+WORKDIR cruise-control
+
+RUN ./gradlew jar copyDependantLibs
```

Comment:
Could we use a two stage build or something to make this part a bit less nasty? (such as having JDK and all build artifacts directly in the image)

Reply:
Updated to mult-stage builds